### PR TITLE
repair: release resources of shard_repair_task_impl

### PR DESCRIPTION
**Pull request by Deexie**
Before integration with task manager the state of one shard repair was kept in repair_info. repair_info object was destroyed immediately after shard repair was finished.

In an integration process repair_info's fields were moved to shard_repair_task_impl as the two served the similar purposes. Though, shard_repair_task_impl isn't immediately destoyed, but is kept in task manager for task_ttl seconds after it's complete. Thus, some of repair_info's fields have their lifetime prolonged, which makes the repair state change delayed.

Release shard_repair_task_impl resources immediately after shard repair is finished.

Fixes: #15505.

### CODE REVIEW

**Comment by Deexie** [comment]
Introduced by https://github.com/scylladb/scylladb/pull/12066.
Affected branches: [scylla-5.3.0-rc0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.3.0-rc0) [scylla-5.2.8](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.8) [scylla-5.2.7](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.7) [scylla-5.2.6](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.6) [scylla-5.2.5](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.5) [scylla-5.2.4](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.4) [scylla-5.2.3](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.3) [scylla-5.2.2](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.2) [scylla-5.2.1](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.1) [scylla-5.2.0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0) [scylla-5.2.0-rc5](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc5) [scylla-5.2.0-rc4](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc4) [scylla-5.2.0-rc3](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc3) [scylla-5.2.0-rc2](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc2) [scylla-5.2.0-rc1](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc1) [scylla-5.2.0-rc0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc0)

**Comment by mykaul** [comment]
> Introduced by #12066. Affected branches: [scylla-5.3.0-rc0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.3.0-rc0) [scylla-5.2.8](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.8) [scylla-5.2.7](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.7) [scylla-5.2.6](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.6) [scylla-5.2.5](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.5) [scylla-5.2.4](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.4) [scylla-5.2.3](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.3) [scylla-5.2.2](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.2) [scylla-5.2.1](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.1) [scylla-5.2.0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0) [scylla-5.2.0-rc5](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc5) [scylla-5.2.0-rc4](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc4) [scylla-5.2.0-rc3](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc3) [scylla-5.2.0-rc2](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc2) [scylla-5.2.0-rc1](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc1) [scylla-5.2.0-rc0](https://github.com/scylladb/scylladb/releases/tag/scylla-5.2.0-rc0)

So I assume we'll need a backport?

**Comment by Deexie** [comment]
> So I assume we'll need a backport?

Yes, I think so, that's quite a cumbersome bug

**Comment by Deexie** [comment]
- do not release `ranges` and `table_ids` which may be used later

**Comment by Deexie** [comment]
> CI state **`ABORTED`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3843/

Looks like #5336 hits again

**Comment by Deexie** [comment]
> CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3866/

 #14315

**Comment by Deexie** [comment]
- rebase

**Comment by bhalevy** [comment]
> > CI state **`FAILURE`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/3866/
> 
> #14315

Does it need to be reopened?

**Comment by Deexie** [comment]
> Does it need to be reopened?

I'm not sure. From the issue:
> I think this issue should be closed unless internode_compression is really broken, i.e., doesn't work at all.

And the test is just flaky, I ran it locally and it passed. Maybe it's worth openning it's own issue as the linked one does not mention the `test_internode_compression_between_datacenters` test?

**Comment by bhalevy** [comment]
> > Does it need to be reopened?
> 
> I'm not sure. From the issue:
> 
> > I think this issue should be closed unless internode_compression is really broken, i.e., doesn't work at all.
> 
> And the test is just flaky, I ran it locally and it passed. Maybe it's worth openning its own issue as the linked one does not mention the `test_internode_compression_between_datacenters` test?

Yes please, so at least we have the paper trail to track it.

**Review by bhalevy — APPROVED** [review]

**Comment by bhalevy** [comment]
> > > Does it need to be reopened?
> > 
> > 
> > I'm not sure. From the issue:
> > > I think this issue should be closed unless internode_compression is really broken, i.e., doesn't work at all.
> > 
> > 
> > And the test is just flaky, I ran it locally and it passed. Maybe it's worth openning its own issue as the linked one does not mention the `test_internode_compression_between_datacenters` test?
> 
> Yes please, so at least we have the paper trail to track it.

I opened a new issue to track this: https://github.com/scylladb/scylladb/issues/15530
since it seems like a recent regression.

File: compaction/compaction_manager.hh [hunk 1]
```diff
@@ -487,7 +487,6 @@ public:
     compaction_task_executor(compaction_task_executor&&) = delete;
     compaction_task_executor(const compaction_task_executor&) = delete;
 
-    virtual void release_resources() noexcept {}
```
Comment (denesb) [inline compaction/compaction_manager.hh:490]: I don't understand these seemingly unrelated changes in `compaction/`.

File: tasks/task_manager.cc [hunk 2]
```diff
@@ -250,6 +250,10 @@ bool task_manager::task::is_complete() const noexcept {
     return _impl->is_complete();
 }
 
+void task_manager::task::release_resources() noexcept {
```
Comment (denesb) [inline tasks/task_manager.cc:253]: Ah, so before only compaction tasks had this method, but now all tasks can have it? That is why `compaction/` is also changed?

File: repair/repair.cc [hunk 3]
```diff
@@ -1390,6 +1404,9 @@ future<> repair::data_sync_repair_task_impl::run() {
                         id, std::move(data_centers), std::move(hosts), std::move(ignore_nodes), reason, hints_batchlog_flushed, ranges_parallelism);
                 task_impl_ptr->neighbors = std::move(neighbors);
                 auto task = co_await local_repair._repair_module->make_task(std::move(task_impl_ptr), parent_data);
+                auto release_task_resources = defer([&] () noexcept {
+                    task->release_resources();
+                });
```
Comment (denesb) [inline repair/repair.cc:1409]: Can the task manager automatically call this when the task is done?

Comment (Deexie) [reply repair/repair.cc:1409]: I was thinking about it and did it like this for one reason:
`release_resources` method is useful mainly for tasks which replace already existing classes (like `shard_repair_task_impl` replaced `repair_info`) and thus the lifetime of their object is prolonged.

For both `shard_repair_task_impl` and `compaction_task_executor`s the original object was destroyed immediately after the operation is finished. But this doesn't need to be the case for all future tasks. Let's say an original object isn't destroyed after operation is done, but a shared_ptr to it is stored in some vector and its destruction is invoked by some other event. It would be nice if we could reuse `release_resources` method also in this case.

I'm not sure how probable this situation is and whether we should consider it at this moment. If you don't think that's a valid concern I will change that.

File: tasks/task_manager.cc (hunk 2 above)
Comment (denesb) [reply tasks/task_manager.cc:253]: Agree, but now that this is a generic thing across tasks, can we invoke it automatically after the task finishes, and before it is entering the TTL phase? I see below this is done manually. Can be a follow-up if this is not simple, no need to delay fixing this bug.

File: repair/repair.cc (hunk 3 above)
Comment (denesb) [reply repair/repair.cc:1409]: I don't have a strong opinion on this. We can change this after we have more tasks and have a perspective on it.
